pkgs/test_core/lib/src/runner/reporter/github.dart

```diff
@@ -0,0 +1,198 @@
+    _subscriptions.add(_engine.onTestStarted.listen(_onTestStarted));
+    _subscriptions.add(_engine.success.asStream().listen(_onDone));
+
+    // Add a spacer between pre-test output and the test results.
+    _sink.writeln();
+  }
+
+  @override
+  void pause() {
+    if (_paused) return;
+    _paused = true;
+
+    for (var subscription in _subscriptions) {
+      subscription.pause();
+    }
+  }
+
+  @override
+  void resume() {
+    if (!_paused) return;
+    _paused = false;
+
+    for (var subscription in _subscriptions) {
+      subscription.resume();
+    }
+  }
+
+  void _cancel() {
+    for (var subscription in _subscriptions) {
+      subscription.cancel();
+    }
+    _subscriptions.clear();
+  }
+
+  /// A callback called when the engine begins running [liveTest].
+  void _onTestStarted(LiveTest liveTest) {
+    // Convert the future to a stream so that the subscription can be paused or
+    // canceled.
+    _subscriptions.add(
+        liveTest.onComplete.asStream().listen((_) => _onComplete(liveTest)));
+
+    // Collect messages from tests as they are emitted.
+    _subscriptions.add(liveTest.onMessage.listen((message) {
+      if (_completedTests.contains(liveTest)) {
+        // The test has already completed and it's previous messages were
+        // written out; ensure this post-completion output is not lost.
+        _sink.writeln(message.text);
+      } else {
+        _testMessages.putIfAbsent(liveTest, () => []).add(message);
+      }
+    }));
+  }
+
+  /// A callback called when [liveTest] finishes running.
+  void _onComplete(LiveTest test) {
+    final errors = test.errors;
+    final messages = _testMessages[test] ?? [];
+    final skipped = test.state.result == Result.skipped;
+    final failed = errors.isNotEmpty;
+    final loadSuite = test.suite is LoadSuite;
+    final synthetic = loadSuite ||
+        test.individualName == setUpAllName ||
+        test.individualName == tearDownAllName;
+
+    // Mark this test as having completed.
+    _completedTests.add(test);
+
+    // Don't emit any info for loadSuite, setUpAll, or tearDownAll tests
+    // unless they contain errors or other info.
+    if (synthetic && (errors.isEmpty && messages.isEmpty)) {
+      return;
+    }
+
+    var defaultIcon =
+        synthetic ? _GithubHelper.synthetic : _GithubHelper.passed;
+    final prefix = failed
+        ? _GithubHelper.failed
+        : skipped
+            ? _GithubHelper.skipped
+            : defaultIcon;
+    final statusSuffix = failed
+        ? ' (failed)'
+        : skipped
+            ? ' (skipped)'
+            : '';
+
+    var name = test.test.name;
+    if (!loadSuite) {
+      if (_printPath && test.suite.path != null) {
+        name = '${test.suite.path}: $name';
+      }
+    }
+    _sink.writeln(_helper.startGroup('$prefix $name$statusSuffix'));
+    // TODO: Note that printing messages and errors in this manner could display
+    // them in a different order than they were generated.
```

Comment:
Do you mean specifically that messages and errors won't be interleaved, not that messages could be out of order relative to other messages?

I think that's probably fine.

I'm curious though if we should try to do anything with tests that print after they are complete - do I understand correctly that they'd get lost?

Reply:
The messages will be in the right order, and the errors will be in the right order, but this might interleave the two types differently than how they'd been created.

Reply:
> I'm curious though if we should try to do anything with tests that print after they are complete - do I understand correctly that they'd get lost?

No - we specifically handle that case (before getting the associated test to pass we had been dropping them). We now track when tests are completed; if we get output for a test which has already completed, we just emit the output directly to the sink (w/o any batching). The output will appear between test groups - it may not be clear what test it's from, but at least it won't be lost.